### PR TITLE
[WebGPU EP] Support GroupQueryAttention

### PR DESCRIPTION
**Pull request by satyajandhyala**
### Description
Add GQA operator support in WebGPU EP.

### Motivation and Context
Enable some LLM models to run using Webgpu EP.

### CODE REVIEW

**Comment by skottmckay** [comment]
Does webgpu_contrib_kernel.cc need an update to uncomment the registration of the op?

>       // BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kWebGpuExecutionProvider, kMSDomain, 1, GroupQueryAttention)>,

**Comment by skottmckay** [comment]
iOS build fails due to CheckInputs being defined in a header leading to multiple instances of it.

```
ld.lld: error: duplicate symbol: onnxruntime::contrib::group_query_attention_helper::CheckInputs(onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, void*, int, int, onnxruntime::Tensor const*, onnxruntime::Tensor const*, float, float)
>>> defined at group_query_attention_helper.h:27 (/Users/scmckay/src/github/ort/onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:27)
>>>            group_query_attention.cc.o:(onnxruntime::contrib::group_query_attention_helper::CheckInputs(onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, void*, int, int, onnxruntime::Tensor const*, onnxruntime::Tensor const*, float, float)) in archive libonnxruntime_providers_webgpu.a
>>> defined at group_query_attention_helper.h:27 (/Users/scmckay/src/github/ort/onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:27)
>>>            group_query_attention.cc.o:(.text._ZN11onnxruntime7contrib28group_query_attention_helper11CheckInputsEPKNS_6TensorES4_S4_S4_S4_S4_S4_PviiS4_S4_ff+0x0) in archive libonnxruntime_providers.a

ld.lld: error: duplicate symbol: onnxruntime::contrib::group_query_attention_helper::CheckInputs(onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, void*, int, int, onnxruntime::Tensor const*, onnxruntime::Tensor const*, float, float, int)
>>> defined at group_query_attention_helper.h:282 (/Users/scmckay/src/github/ort/onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:282)
>>>            group_query_attention.cc.o:(onnxruntime::contrib::group_query_attention_helper::CheckInputs(onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, onnxruntime::Tensor const*, void*, int, int, onnxruntime::Tensor const*, onnxruntime::Tensor const*, float, float, int)) in archive libonnxruntime_providers_webgpu.a
>>> defined at group_query_attention_helper.h:282 (/Users/scmckay/src/github/ort/onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:282)
>>>            group_query_attention.cc.o:(.text._ZN11onnxruntime7contrib28group_query_attention_helper11CheckInputsEPKNS_6TensorES4_S4_S4_S4_S4_S4_PviiS4_S4_ffi+0x0) in archive libonnxruntime_providers.a
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: onnxruntime/contrib_ops/webgpu/bert/attention.cc [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+  } else {
+    shader.MainFunctionBody() << ";\n";
+  }
+
+  return Status::OK();
+}
+
+Status TransferBSDToBNSH(onnxruntime::webgpu::ComputeContext& context, int num_heads, int sequence_length,
+                         int head_size, const Tensor* input_tensor, const Tensor* bias, int bias_offset, Tensor* output_tensor) {
+  assert(input_tensor->Shape().GetDims().size() == 3);
```
Comment (fs-eire) [inline onnxruntime/contrib_ops/webgpu/bert/attention.cc:45]: assert -> ORT_ENFORCE

File: onnxruntime/contrib_ops/webgpu/bert/attention.cc [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+  if (has_bias) {
+    program.AddInput({bias, ProgramTensorMetadataDependency::TypeAndRank});
+  }
+
+  return context.RunProgram(program);
+};
+
+std::string InitVarStub(const Tensor* seqlens_k, const Tensor* total_seqlen_tensor, bool init_past_sequence_length) {
+  std::ostringstream ss;
```
Comment (fs-eire) [inline onnxruntime/contrib_ops/webgpu/bert/attention.cc:72]: use macro `SS` with an estimated length (defined in string_macros.h)

File: onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h [hunk 3]
```diff
@@ -11,18 +11,19 @@ namespace onnxruntime {
...
 namespace group_query_attention_helper {
 
-Status CheckInputs(const Tensor* query,
-                   const Tensor* key,
-                   const Tensor* value,
-                   const Tensor* past_key,
-                   const Tensor* past_value,
-                   const Tensor* cos_cache,
-                   const Tensor* sin_cache,
+template <typename T = Tensor>
```
Comment (fs-eire) [inline onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:14]: template seems not necessary

a more straightforward way is to keep delaration only in .h, and put the implementation in .cc

Comment (satyajandhyala) [reply onnxruntime/contrib_ops/cpu/bert/group_query_attention_helper.h:14]: That is true. However doing so requires modifying the code, moving definition in a new source file and keep the declaration in the header file. It is not a generally good practice to put code like this in the header file. I am trying to work with minimal changes. multihead_attention_helper.h works the same way.

**Comment by satyajandhyala** [comment]
closing this pr because I opened the same code change targeting main instead of fs-eire/webgpu-ep

**Pull request closed without merging**